yapapi/services.py

```diff
@@ -191,26 +204,143 @@ def _respond_nowait(self, message: Optional[Any], response_to: Optional[ServiceS
         self._ctx.start()
         yield self._ctx.commit()
 
-    async def run(self):
-        """Implement the `running` state of the service."""
+    async def run(self) -> AsyncGenerator[WorkItem, Awaitable[List[events.CommandEvent]]]:
+        """Implement the handler for the `running` state of the service.
+
+        To be overridden by the author of a specific Service class.
+
+        Should contain any operations needed ensure the continual operation of a service.
+
+        As a handler, it's expected to be a generator that yields `WorkItems` (generated using the
+        service's instance of the work context) that are then dispatched to the activity by the
+        engine.
+
+        Results of those batches can then be captured by awaiting the values captured from yield
+        statements.
+
+        ### Example
+
+        ```
+        async def run(self):
+            while True:
+                self._ctx.run("/golem/run/report", "--stats")  # index 0
+                future_results = yield self._ctx.commit()
+                results = await future_results
+                stats = results[0].stdout.strip()  # retrieve from index 0
+                print(f"stats: {stats}")
+        ```
+
+        ### Default implementation
+
+        Because the nature of the operations required during the "running" state depends directly
+        on the specifics of a given Service and because it's entirely plausible for a service
+        not to require any direct interaction with the exe-unit (runtime) from the requestor's end
+        after the service has been started, the default is to just wait indefinitely without
+        producing any batches.
+        """
 
         await asyncio.Future()
-        yield
+        yield  # type: ignore # unreachable because of the indefinite wait above
+
+    async def shutdown(self) -> AsyncGenerator[WorkItem, Awaitable[List[events.CommandEvent]]]:
+        """Implement the handler for the `stopping` state of the service.
 
-    async def shutdown(self):
-        """Implement the `stopping` state of the service."""
+        To be overridden by the author of a specific Service class.
+
+        Should contain any operations that the requestor needs to ensure the instance is correctly
+        shut-down - e.g. that its final state is retrieved.
+
+        As a handler, it's expected to be a generator that yields `WorkItems` (generated using the
+        service's instance of the work context) that are then dispatched to the activity by the
+        engine.
+
+        Results of those batches can then be captured by awaiting the values captured from yield
+        statements.
+
+        This handler will only be called if the activity running the service is still available.
+        If the activity has already been deemed terminated or if the connection with the provider
+        has been lost, the service will transition to the `terminated` state and the shutdown
+        handler won't be run.
+
```

Comment:
What is the `self._ctx.terminate()` here? Is it required in the `shutdown()` handler?
What would happen if I create a `shutdown` handler without it?

I think this requires some explanation (similar to `ctx.deploy()/ctx.start()` in `start()`).

Reply:
nothing is required, it's runtime-dependent ... added a minor additional comment but I'm unsure what more needs to be said here